memory-db/src/lib.rs

```diff
@@ -286,31 +307,37 @@ where
 		match self.data.entry(key) {
 			Entry::Occupied(mut entry) =>
 				if entry.get().1 == 1 {
-					Some(entry.remove().0)
+					let (value, _) = entry.remove();
+					self.malloc_callback.on_value_removed(&value);
+					Some(value)
 				} else {
 					entry.get_mut().1 -= 1;
 					None
 				},
 			Entry::Vacant(entry) => {
-				entry.insert((T::default(), -1)); // FIXME: shouldn't it be purged?
+				let value = T::default();
+				self.malloc_callback.on_value_inserted(&value);
+				entry.insert((value, -1)); // FIXME: shouldn't it be purged?
```

Comment:
is this FIXME still relevant?

Reply:
I do not have much context on it, but I would expect it to work this way: removing reach 0 so we purge (0 is default for any key so it is unneeded info), removing on non existing is additional info so we add it (that way next insert is invalidated).
So I would thing removing comment would be fine.
I think it (-1) should only happen in case where there was concurrency, and we inject resulting deltas in a different order than processing.